### PR TITLE
feature (issue 268): fixed deserializer test

### PR DESCRIPTION
**Pull request by gv-timur**
fixed deserializer test
deserializers refactoring

### CODE REVIEW

**Comment by Mingela** [comment]
I'd say that's okay to start with, what do you think @arndey? though let's add more within the PR first based on other genesis blocks as you mentioned

File: modules/block/src/test/kotlin/jp/co/soramitsu/iroha2/DeserializerTest.kt [hunk 1]
```diff
@@ -14,6 +13,10 @@ class DeserializerTest {
         val block = JSON_SERDE.convertValue(node, RawGenesisBlock::class.java)
 
         assert(block.transactions.isNotEmpty())
-        assert(block.transactions.first().isi.size == 4)
+        assert(block.transactions.first().isi.size == 7)
```
Comment (arndey) [inline modules/block/src/test/kotlin/jp/co/soramitsu/iroha2/DeserializerTest.kt:16]: left a comment why 7

**Review by arndey — APPROVED** [review]

File: modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt [hunk 2]
```diff
@@ -165,13 +181,81 @@ private fun sealedDeserializeGrantBox(p: JsonParser, mapper: ObjectMapper): Gran
     )
 }
 
-private fun sealedDeserializeValue(p: JsonParser): Value {
+private fun sealedDeserializeValue(p: JsonParser, mapper: ObjectMapper): Value {
     val node = p.readValueAsTree<JsonNode>().fields().next()
-    return node.value.asText().asValue()
+    val param = if (node.key.contains("Id")) "Id" else node.key
+
+    val clazz = when (param) {
```
Comment (arndey) [inline modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt:188]: deserves to be executed in a separate private method

File: modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt [hunk 3]
```diff
@@ -165,13 +181,81 @@ private fun sealedDeserializeGrantBox(p: JsonParser, mapper: ObjectMapper): Gran
...
+        val subtype = clazz.nestedClasses.find { clazz ->
+            !clazz.isCompanion && clazz.simpleName == name
+        } ?: throw DeserializationException("Class with constructor($param) not found")
+
+        val argTypeName = subtype.primaryConstructor?.parameters
+            ?.firstOrNull()?.type?.toString()
+            ?: throw DeserializationException("Subtype parameter not found by $param")
+
+        val arg = mapper.convertValue(value, argTypeName.asClass())
+        return when (clazz) {
```
Comment (arndey) [inline modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt:232]: and this

File: modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt [hunk 4]
```diff
@@ -103,55 +132,249 @@ val JSON_SERDE by lazy {
...
+    val param = node.key.removePrefix("New")
+    val subtype = RegistrableBox::class.nestedClasses.find { clazz ->
+        !clazz.isCompanion && clazz.simpleName == param
+    }
+    val argTypeName = subtype?.primaryConstructor?.parameters
+        ?.firstOrNull()?.type?.toString()
+        ?: throw DeserializationException("Subtype parameter not found by $param")
+
+    val arg = mapper.convertValue(node.value, argTypeName.asClass())
+    return when (arg) {
```
Comment (arndey) [inline modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt:285]: and this

File: modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt [hunk 5]
```diff
@@ -103,55 +132,249 @@ val JSON_SERDE by lazy {
...
+
+    while (iter.hasNext()) {
+        val node = iter.next()
+        nodes.add(node)
+    }
+    val definitionId = jsonNode.fields().next()
+    val params = nodes[1].fields()
+    val tokenMap = mutableMapOf<Name, Value>()
+
+    val arg = mapper.convertValue(definitionId.value, "jp.co.soramitsu.iroha2.generated.datamodel.permission.token.TokenId".asClass()) as TokenId
```
Comment (arndey) [inline modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Serde.kt:352]: arg, arg2, arg3 -> smth like tokenId, name, value

File: modules/block/src/test/kotlin/jp/co/soramitsu/iroha2/DeserializerTest.kt [hunk 6]
```diff
@@ -1,19 +1,56 @@
...
+        val genesis = Genesis(block)
+        val newJson = genesis.asJson()
+            .replace("\r\n", "")
+            .replace(" ", "")
+            .replace(regex, "publicKey")
+        val initialJson = json.readText()
+            .replace("\r\n", "")
+            .replace(" ", "")
+            .replace(regex, "publicKey")
+//        assertEquals(initialJson, newJson)
```
Comment (arndey) [inline modules/block/src/test/kotlin/jp/co/soramitsu/iroha2/DeserializerTest.kt:28]: ?

File: modules/block/src/test/kotlin/jp/co/soramitsu/iroha2/DeserializerTest.kt [hunk 7]
```diff
@@ -1,19 +1,56 @@
...
```
Comment (arndey) [inline modules/block/src/test/kotlin/jp/co/soramitsu/iroha2/DeserializerTest.kt:52]: ?

**Review by arndey — CHANGES_REQUESTED** [review]
some comments

**Review by Mingela — APPROVED** [review]

**Review by arndey — APPROVED** [review]

**Pull request merged**